### DISCUSSION TITLE
Adding a new platform (z/OS) to the CI

### DISCUSSION THREAD

**Discussion by IgorTodorovskiIBM** [curl/curl#15733] [General]
Hi, I work on the IBM z/OS platform and Curl builds on the platform (I made a contribution previously for it here: https://github.com/curl/curl/pull/11241)

I am curious what it would take to add the z/OS platform to the CI. Are there any particular requirements for enabling CI on another platform? Is there an example PR?

**Comment by dfandrich** [comment]
There are a few ways to do it. The most straightforward is to add an autobuild
using the instructions at https://curl.se/dev/howto.html  This way, you have
control over what you build and when, then the results are uploaded and
displayed at https://curl.se/dev/builds.html  This way is simple with few
requirements, but the big downside is that this method only checks changes
after they're submitted, so failures in PRs won't be noticed until after
submission.  The results are also only available to people who check the
dashboard manually.

The more useful way is to integrate it into the GitHub CI pipeline, and there
are a few ways to do this as well. The easier one would be to configure a
GitHub self-hosted runner which would operate much like any other GitHub
Actions build but on your hardware instead of GitHub's. It's extremely likely
that the necessary software isn't compatible with z/OS so this probably isn't
an option for you, even if it would probably otherwise be one of the easiest
and best integrated.

Another is to implement the same hooks as CI services like Appveyor, Azure,
Circle CI, Cirrus, GitHub Actions and Travis CI do and become a first-class
GitHub-supported CI service. I don't know all that this entails, but may
involve creating a GitHub App for the best integration, so it's probably a huge
job. However, this gives the best developer experience as code in PRs is built
and any failures are automatically reflected in the PR summary page and
e-mailed to developers.

Alternately, a basic CI can be implemented by installing a GitHub webhook to
trigger a build on your build server on each new PR with the results updating
the PR status. Someone (Dago) has done this for curl for Solaris builds using
Buildbot, but I haven't seen those results in PRs lately so I don't know the
current status of that. You can still see daily builds from that system in the
autobuild results page, though. If Buildbot or a similar system is available on
z/OS, this approach is probably going to give the best bang for the buck.
